<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta http-equiv="X-UA-Compatible" content="IE=edge"/><title>Tutorial · Test Site</title><meta name="viewport" content="width=device-width"/><meta name="generator" content="Docusaurus"/><meta name="description" content="Following the steps on this page you will set up an example InterMine."/><meta name="docsearch:version" content="next"/><meta name="docsearch:language" content="en"/><meta property="og:title" content="Tutorial · Test Site"/><meta property="og:type" content="website"/><meta property="og:url" content="https://egbobright.github.io/test-site/"/><meta property="og:description" content="Following the steps on this page you will set up an example InterMine."/><meta property="og:image" content="https://egbobright.github.io/test-site/img/undraw_online.svg"/><meta name="twitter:card" content="summary"/><meta name="twitter:image" content="https://egbobright.github.io/test-site/img/undraw_tweetstorm.svg"/><link rel="shortcut icon" href="/test-site/img/favicon.ico"/><link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css"/><link rel="alternate" type="application/atom+xml" href="https://egbobright.github.io/test-site/blog/atom.xml" title="Test Site Blog ATOM Feed"/><link rel="alternate" type="application/rss+xml" href="https://egbobright.github.io/test-site/blog/feed.xml" title="Test Site Blog RSS Feed"/><script type="text/javascript" src="https://buttons.github.io/buttons.js"></script><script src="/test-site/js/scrollSpy.js"></script><link rel="stylesheet" href="/test-site/css/main.css"/><script src="/test-site/js/codetabs.js"></script></head><body class="sideNavVisible separateOnPageNav"><div class="fixedHeaderContainer"><div class="headerWrapper wrapper"><header><a href="/test-site/"><img class="logo" src="/test-site/img/favicon.ico" alt="Test Site"/><h2 class="headerTitleWithLogo">Test Site</h2></a><a href="/test-site/versions"><h3>next</h3></a><div class="navigationWrapper navigationSlider"><nav class="slidingNav"><ul class="nav-site nav-site-internal"><li class="siteNavGroupActive siteNavItemActive"><a href="/test-site/docs/next/doc4" target="_self">Docs</a></li><li class="siteNavGroupActive"><a href="/test-site/docs/next/doc1" target="_self">API</a></li><li class=""><a href="/test-site/help" target="_self">Help</a></li><li class=""><a href="/test-site/blog/" target="_self">Blog</a></li></ul></nav></div></header></div></div><div class="navPusher"><div class="docMainWrapper wrapper"><div class="docsNavContainer" id="docsNav"><nav class="toc"><div class="toggleNav"><section class="navWrapper wrapper"><div class="navBreadcrumb wrapper"><div class="navToggle" id="navToggler"><div class="hamburger-menu"><div class="line1"></div><div class="line2"></div><div class="line3"></div></div></div><h2><i>›</i><span>Docusaurus</span></h2><div class="tocToggler" id="tocToggler"><i class="icon-toc"></i></div></div><div class="navGroups"><div class="navGroup"><h3 class="navGroupCategoryTitle">Docusaurus</h3><ul class=""><li class="navListItem"><a class="navItem" href="/test-site/docs/next/doc1">Example Page</a></li><li class="navListItem navListItemActive"><a class="navItem" href="/test-site/docs/next/doc4">Tutorial</a></li></ul></div><div class="navGroup"><h3 class="navGroupCategoryTitle">First Category</h3><ul class=""><li class="navListItem"><a class="navItem" href="/test-site/docs/next/doc2">document number 2</a></li></ul></div><div class="navGroup"><h3 class="navGroupCategoryTitle">Second Category</h3><ul class=""><li class="navListItem"><a class="navItem" href="/test-site/docs/next/doc3">This is document number 3</a></li></ul></div></div></section></div><script>
            var coll = document.getElementsByClassName('collapsible');
            var checkActiveCategory = true;
            for (var i = 0; i < coll.length; i++) {
              var links = coll[i].nextElementSibling.getElementsByTagName('*');
              if (checkActiveCategory){
                for (var j = 0; j < links.length; j++) {
                  if (links[j].classList.contains('navListItemActive')){
                    coll[i].nextElementSibling.classList.toggle('hide');
                    coll[i].childNodes[1].classList.toggle('rotate');
                    checkActiveCategory = false;
                    break;
                  }
                }
              }

              coll[i].addEventListener('click', function() {
                var arrow = this.childNodes[1];
                arrow.classList.toggle('rotate');
                var content = this.nextElementSibling;
                content.classList.toggle('hide');
              });
            }

            document.addEventListener('DOMContentLoaded', function() {
              createToggler('#navToggler', '#docsNav', 'docsSliderActive');
              createToggler('#tocToggler', 'body', 'tocActive');

              var headings = document.querySelector('.toc-headings');
              headings && headings.addEventListener('click', function(event) {
                var el = event.target;
                while(el !== headings){
                  if (el.tagName === 'A') {
                    document.body.classList.remove('tocActive');
                    break;
                  } else{
                    el = el.parentNode;
                  }
                }
              }, false);

              function createToggler(togglerSelector, targetSelector, className) {
                var toggler = document.querySelector(togglerSelector);
                var target = document.querySelector(targetSelector);

                if (!toggler) {
                  return;
                }

                toggler.onclick = function(event) {
                  event.preventDefault();

                  target.classList.toggle(className);
                };
              }
            });
        </script></nav></div><div class="container mainContainer docsContainer"><div class="wrapper"><div class="post"><header class="postHeader"><h1 id="__docusaurus" class="postHeaderTitle">Tutorial</h1></header><article><div><span><p>Following the steps on this page you will set up an example InterMine.
You will:</p>
<blockquote>
<ul>
<li>Load some real data sets for Malaria (<em>P. falciparum</em>)</li>
<li>Learn about how data integration works</li>
<li>Deploy a webapp to query the data</li>
</ul>
</blockquote>
<hr>
<p><strong>NOTE</strong></p>
<p>See <a href="'https://intermine.readthedocs.io/en/get-started/quick-start'">Quick start</a> if you are impatient and just want to run an InterMine.</p>
<hr>
<h1><a class="anchor" aria-hidden="true" id="getting-started"></a><a href="#getting-started" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Getting Started</h1>
<h2><a class="anchor" aria-hidden="true" id="software"></a><a href="#software" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Software</h2>
<p>We use <a href="http://git-scm.com">git</a> to manage and distribute source code
and <a href="http://gradle.org">gradle</a> as our build system. For this tutorial
you will need the following software packages installed locally and
running:</p>
<ul>
<li>PostgreSQL</li>
<li>Git</li>
<li>Java</li>
<li>Tomcat</li>
<li>Solr</li>
<li>Perl (for the final build script)</li>
</ul>
<p>See <a href="'https://intermine.readthedocs.io/en/latest/system-requirements/software/'">Software</a> for configuration details.</p>
<h2><a class="anchor" aria-hidden="true" id="biotestmine"></a><a href="#biotestmine" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>BioTestMine</h2>
<p>Download the mine code from GitHub.</p>
<hr>
<p><code>$ mkdir git</code>
<code>$ cd git</code>
<code>~/git $ git clone https://github.com/intermine/biotestmine</code></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="get-rid-of-daemons"></a><a href="#get-rid-of-daemons" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Get rid of daemons</h3>
<p>Gradle has helper processes enabled by default. We're going to disable
those by setting <code>-Dorg.gradle.daemon=false</code></p>
<hr>
<p><code>~/git $ export GRADLE_OPTS=&quot;-Dorg.gradle.daemon=false&quot;</code></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="help-somethings-gone-wrong"></a><a href="#help-somethings-gone-wrong" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Help! Something's gone wrong.</h3>
<p>If at any point you need help or have a quick (or not so quick)
question, please <a href="http://intermine.org/contact/">get in touch</a>! We have
a discord server, twitter and a developer mailing list.</p>
<h1><a class="anchor" aria-hidden="true" id="biotest-mine"></a><a href="#biotest-mine" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>BioTest Mine</h1>
<p>BioTestMine is a dummy test mine we use to test out new features which
contains real (old) data for Malaria (<em>P. falciparum</em>).</p>
<p>To get started, change into the directory you checked out the
BiotestMine source code to and look at the sub-directories:</p>
<hr>
<p><code>~/git $ cd biotestmine</code>
<code>~/git/biotestmine $ ls</code></p>
<hr>
<table>
<thead>
<tr><th>directory/file</th><th>purpose</th></tr>
</thead>
<tbody>
<tr><td>/dbmodel</td><td>contains information about the data model and related configuration files</td></tr>
<tr><td>/webapp</td><td>basic configuration for the webapp</td></tr>
<tr><td>/data</td><td>contains a tar file with data to load</td></tr>
<tr><td>build.gradle</td><td>The --stacktrace option will display complete error messages if there is a problem.</td></tr>
<tr><td>gradle.properties</td><td>Sets system variables. Determines which version of InterMine you use.</td></tr>
<tr><td>settings.gradle</td><td>Sets gradle projects. Do not edit.</td></tr>
<tr><td>project.xml</td><td>Configures which data parsers are run during your build.</td></tr>
</tbody>
</table>
<p>There is also a gradle directory (/gradle) and executables (gradlew, gradle.bat).</p>
<h2><a class="anchor" aria-hidden="true" id="projectxml"></a><a href="#projectxml" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Project.xml</h2>
<p>The project.xml allows you to configure which data to load into your Mine. The file has two sections: sources and post-processing.
<sources></p>
<p>The <source> elements list and configure the data sources to be loaded. A source can have a name and a type.
type
Corresponds to the name of the bio-source artifact (jar) which includes parsers to retrieve data and information on how it will be integrated.
name
can be anything and can be the same as type, using a more specific name allows you to define specific integration keys.</p>
<p><source> elements can have several properties depending on source type: src.data.dir, src.data.file and src.data.includes are all used to define locations of files that the source should load. Other properties are used as parameters to specific parsers.
&lt;post-processing&gt;</p>
<p>Specific operations can be performed on the Mine once data is loaded, these are listed here as &lt;post-process&gt; elements. We will look at these in more detail later.
Data to load</p>
<p>The biotestmine git repository includes a tar file with data to load into BiotestMine. These are real, complete data sets for P. falciparum (but very old!).</p>
<p>We will load genome annotation from PlasmoDB, protein data from UniProt and GO annotation also from PlasmoDB.</p>
<p>See <a href="'https://intermine.readthedocs.io/en/latest/get-started/tutorial/test-data/'">Data files to integrate</a> for details on the data.
Copy this to a local directory (your home directory is fine for this workshop) and extract the archive:</p>
<hr>
<p><code>$ cd</code>
<code>$ cp git/biotestmine/data/malaria-data.tar.gz .</code>
<code>$ tar -zxvf malaria-data.tar.gz</code></p>
<hr>
<p>A dummy project XML file is available in the /data/ directory. Copy it into your biotestmine directory, then edit project.xml to point each source at the extracted data, just replace /data with /home/username (or on a mac /Users/username). Do use the absolute path.</p>
<hr>
<p><code>$ cd ~/git/biotestmine</code>
<code>$ cp ~/git/biotestmine/data/project.xml .</code>
<code>~/git/biotestmine $ sed -i 's/\DATA\_DIR/\/home\/username/g' project.xml</code></p>
<hr>
<p>For example, the uniprot-malaria source:</p>
<pre><code class="hljs"> &lt;sources&gt;
  &lt;source <span class="hljs-type">name</span>="uniprot-malaria" <span class="hljs-keyword">type</span>="uniprot"&gt;
    &lt;property <span class="hljs-type">name</span>="uniprot.organisms" <span class="hljs-keyword">value</span>="36329"/&gt;
    &lt;property <span class="hljs-type">name</span>="src.data.dir" <span class="hljs-keyword">location</span>="/home/username/malaria/uniprot/"/&gt;
  &lt;/source&gt;
 `` 
</code></pre>
<hr>
<p><strong>NOTE</strong>
All file locations must be absolute not relative paths.</p>
<hr>
<p>The project.xml file is now ready to use.</p>
<h3><a class="anchor" aria-hidden="true" id="properties-file"></a><a href="#properties-file" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Properties file</h3>
<p>Configuration of local databases and tomcat deployment is kept in a MINE_NAME.properties file in a .intermine directory under your home directory. We need to set up a biotestmine.properties file.</p>
<p>If you don’t already have a .intermine directory in your home directory, create one now:</p>
<hr>
<p><code>$ cd</code>
<code>$ mkdir .intermine</code></p>
<hr>
<p>There is a partially completed properties file for BioTestMine already. Copy it into your .intermine directory:</p>
<hr>
<p><code>$ cd</code>
<code>$ cp git/biotestmine/data/biotestmine.properties .intermine/</code></p>
<hr>
<p>Update this properties file with your postgres server location, username and password information for the two databases you just created. The rest of the information is needed for the webapp and will be updated later.</p>
<p>For the moment you need to change PSQL_USER and PSQL_PWD in the db.production and db.common-tgt-items properties.</p>
<hr>
<!---Access to the postgres database to build into and access from the webapp-->  
`db.production.datasource.serverName=localhost`
<!--- port: uncomment the next line if use different port other than 5432-->
<!---`db.production.datasource.port=PORT_NUMBER`-->
`db.production.datasource.databaseName=biotestmine`
`db.production.datasource.user=PSQL_USER`
`db.production.datasource.password=PSQL_PWD`
_______
<p>If you don’t have a password for your postgres account you can leave password blank.</p>
<h3><a class="anchor" aria-hidden="true" id="create-databases"></a><a href="#create-databases" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Create databases</h3>
<p>Finally, we need to create biotestmine and items-biotestmine postgres databases as specified in the biotestmine.properties file:</p>
<hr>
<pre><code class="hljs">$ createdb biotestmine
$ createdb items-biotestmine
</code></pre>
<hr>
<p>New postgres databases default to UTF-8 as the character encoding. This will work with InterMine but performance is better with SQL_ASCII.</p>
<h2><a class="anchor" aria-hidden="true" id="the-data-model"></a><a href="#the-data-model" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>The Data Model</h2>
<p>Now we’re ready to set up a database schema and load some data into our BioTestMine, first some information on how data models are defined in InterMine.</p>
<h3><a class="anchor" aria-hidden="true" id="defining-the-model"></a><a href="#defining-the-model" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Defining the model</h3>
<p>InterMine uses an object-oriented data model, classes in the model and relationships between them are defined in an XML file. Depending on which data types you include you will need different classes and fields in the model, so the model is generated from a core model XML file and any number of additions files. These additions files can define extra classes and fields to be added to the model.</p>
<ul>
<li>Elements of the model are represented by Java classes and references
between them.</li>
<li>These Java classes map automatically to tables in the database
schema.</li>
<li>The object model is defined as an XML file,that defines classes, their attributes and references between classes.</li>
<li>The Java classes and database schema are automatically generated
from an XML file.</li>
</ul>
<p>The model is generated from a core model XML file and any number of
additions files defined in the <a href="https://github.com/intermine/biotestmine/blob/master/dbmodel/build.gradle#L37">dbmodel/build.gradle</a> file.</p>
<h3><a class="anchor" aria-hidden="true" id="corexml"></a><a href="#corexml" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>core.xml</h3>
<p>The core InterMine data model is defined in
<a href="https://github.com/intermine/intermine/blob/master/bio/model/core.xml">core.xml</a>
file.</p>
<p>Note the fields defined for <em>Protein</em>:</p>
<pre><code class="hljs">&lt;<span class="hljs-class"><span class="hljs-keyword">class</span> <span class="hljs-title">name=</span>"<span class="hljs-title">Protein</span>" <span class="hljs-keyword">extends</span></span>=<span class="hljs-string">"BioEntity"</span> is-interface=<span class="hljs-string">"true"</span>&gt;
  &lt;attribute name=<span class="hljs-string">"md5checksum"</span> <span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"java.lang.String"</span>/&gt;
  &lt;attribute name=<span class="hljs-string">"primaryAccession"</span> <span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"java.lang.String"</span>/&gt;
  &lt;attribute name=<span class="hljs-string">"length"</span> <span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"java.lang.Integer"</span>/&gt;
  &lt;attribute name=<span class="hljs-string">"molecularWeight"</span> <span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"java.lang.Double"</span>/&gt;
  &lt;reference name=<span class="hljs-string">"sequence"</span> referenced-<span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"Sequence"</span>/&gt;
  &lt;collection name=<span class="hljs-string">"genes"</span> referenced-<span class="hljs-class"><span class="hljs-keyword">type</span></span>=<span class="hljs-string">"Gene"</span> reverse-reference=<span class="hljs-string">"proteins"</span>/&gt;
&lt;/<span class="hljs-class"><span class="hljs-keyword">class</span><span class="hljs-title">&gt;</span></span>
</code></pre>
<p>Protein is a subclass of [BioEntity]{.title-ref}, defined by
[extends=&quot;BioEntity&quot;]{.title-ref}. The [Protein]{.title-ref} class
will therefore also inherit all fields of [BioEntity]{.title-ref}.</p>
<pre><code class="hljs css language-{.sourceCode .xml}">&lt;class <span class="hljs-attribute">name</span>=<span class="hljs-string">"BioEntity"</span> <span class="hljs-attribute">is-interface</span>=<span class="hljs-string">"true"</span>&gt;
  &lt;attribute <span class="hljs-attribute">name</span>=<span class="hljs-string">"primaryIdentifier"</span> <span class="hljs-attribute">type</span>=<span class="hljs-string">"java.lang.String"</span>/&gt;
  &lt;attribute <span class="hljs-attribute">name</span>=<span class="hljs-string">"secondaryIdentifier"</span> <span class="hljs-attribute">type</span>=<span class="hljs-string">"java.lang.String"</span>/&gt;
<span class="hljs-built_in">..</span>.
</code></pre>
<h3><a class="anchor" aria-hidden="true" id="sequence-ontology"></a><a href="#sequence-ontology" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Sequence Ontology</h3>
<pre><code class="hljs">mineDBModelConfig {
  modelName = &quot;genomic&quot;
  extraModelsStart = &quot;so_additions.xml genomic_additions.xml&quot;
  extraModelsEnd = &quot;&quot;
}
</code></pre>
<p>The first file merged into the core model is the so_additions.xml file. This XML file is generated in the dbmodel/build directory from terms listed in the so_terms file, as configured in <a href="https://github.com/intermine/biotestmine/blob/master/dbmodel/build.gradle#L31">dbmodel/build.gradle</a> file.</p>
<pre><code class="hljs">dbModelConfig {
  soTermListFilePath = &quot;dbmodel/resources/so_terms&quot;
  soAdditionFilePath = &quot;dbmodel/build/so_additions.xml&quot;
}
</code></pre>
<p>The build system creates classes corresponding to the Sequence Ontology
terms.</p>
<h3><a class="anchor" aria-hidden="true" id="additions-files"></a><a href="#additions-files" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Additions files</h3>
<p>The model is then combined with any extra classes and fields defined in the sources to integrate, those listed as <source> elements in project.xml. Look at the <a href="https://github.com/intermine/intermine/blob/master/bio/sources/uniprot/src/main/resources/uniprot_additions.xml">additions file</a> for the UniProt source, for example. This defines extra fields for the Protein class which will be added to those from the core model.</p>
<h3><a class="anchor" aria-hidden="true" id="creating-a-database"></a><a href="#creating-a-database" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Creating a database</h3>
<p>Now run the gradle task to merge all the model components, generate Java
classes and create the database schema:</p>
<hr>
<pre><code class="hljs">&lt;!--- creates the empty database tables --&gt;
</code></pre>
<p><code>~/git/biotestmine $ ./gradlew buildDB</code></p>
<hr>
<p>The clean task is necessary when you have run the task before, it removes the build directory and any previously generated models.</p>
<p>This task has done several things:</p>
<ol>
<li>Merged the core model with other model additions and created a new
XML file:
<code>~/git/biotestmine $ less dbmodel/build/resources/main/genomic_model.xml</code></li>
</ol>
<p>Look for the Protein class, you can see it combines fields from the core model and the UniProt additions file.</p>
<ol start="2">
<li>The so_additions.xml file has also been created using the sequence ontology terms in so_term:</li>
</ol>
<hr>
<p><code>~/git/biotestmine $ less dbmodel/build/so_additions.xml</code></p>
<hr>
<p>Each term from so_term was added to the model, according to the sequence ontology.</p>
<ol start="3">
<li>Generated and compiled a Java class for each of the <class> elements in the     file. For example Protein.java:</li>
</ol>
<hr>
<p><code>~/git/biotestmine $ less dbmodel/build/gen/org/intermine/model/bio/Protein.java</code></p>
<hr>
<p>Each of the fields has appropriate getters and setters generated for it, note that these are interfaces and are turned into actual classes dynamically at runtime - this is how the model copes with multiple inheritance.</p>
<ol start="4">
<li>Automatically created database tables in the postgres database specified in biotestmine.properties as db.production - in our case biotestmine. Log into this database and list the tables and the columns in the protein table:</li>
</ol>
<hr>
<pre><code class="hljs">$ psql biotestmine
biotestmine=#  \d
biotestmine=#  \d protein
</code></pre>
<hr>
<p>The different elements of the model XML file are handled as follows:</p>
<blockquote>
<p><strong><em>attributes</em></strong>
-- there is one column for each attribute of Protein - e.g. primaryIdentifer and length.
<strong><em>references</em></strong>
-- references to other classes are foreign keys to another table - e.g. Protein has a reference called organism to the Organism class so in the database the protein table has a column organismid which would contain an id that appears in the organism table.
<strong><em>collections</em></strong>
-- indirection tables are created for many-to-many collections - e.g. Protein has a collection of Gene objects so an indirection table called genesproteins is created.</p>
</blockquote>
<p>This has also created necessary indexes on the tables:</p>
<pre><code class="hljs">    <span class="hljs-attribute">biotestmine</span>=#  \d genesproteins
</code></pre>
<hr>
<p><strong>WARNING</strong>
Running buildDB will destroy any existing data loaded in the biotestmine database and re-create all the tables.</p>
<hr>
<p>The model XML file is stored in the database once created, this and some other configuration files are held in the intermine_metadata table which has key and value columns:</p>
<pre><code class="hljs">    biotestmine=<span class="hljs-comment">&lt;!-- select key from intermine_metadata--&gt;</span>;
</code></pre>
<h3><a class="anchor" aria-hidden="true" id="loading-data"></a><a href="#loading-data" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Loading Data</h3>
<p>Now we have the correct data model and the correct empty tables in the database. We can now run several data parsers to load our data into our database.</p>
<p>For this tutorial we will run several data integration and post-processing steps manually. This is a good way to learn how the system works and to test individual stages. For running actual builds there is a project_build script that will run all steps specified in project.xml automatically. We will cover this later.</p>
<h3><a class="anchor" aria-hidden="true" id="loading-data-from-a-source"></a><a href="#loading-data-from-a-source" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Loading data from a source</h3>
<p>Loading of data is done by running the integrate gradle task.</p>
<hr>
<pre><code class="hljs">&lt;!--- load the uniprot data source --&gt;
</code></pre>
<p><code>~/git/biotestmine $ ./gradlew integrate -Psource=uniprot-malaria --stacktrace</code></p>
<hr>
<table>
<thead>
<tr><th></th><th></th></tr>
</thead>
<tbody>
<tr><td>./gradlew</td><td>Use the provided gradle wrapper so that we can be sure everyone is using the same version.</td></tr>
<tr><td>integrate</td><td>Gradle task to run the specified data source</td></tr>
<tr><td>-Psource=</td><td>Data source to run. Source name should match the value in your project XML file</td></tr>
<tr><td>--stacktrace</td><td>The –stacktrace option will display complete error messages if there is a problem.</td></tr>
</tbody>
</table>
<p>This will take a couple of minutes to complete, the command runs the
following steps:</p>
<ol>
<li>Checks that a source with name uniprot-malaria exists in project.xml</li>
<li>Reads the UniProt XML files at the location specified by src.data.dir in the project.xml file</li>
<li>Runs the parser included in the UniProt JAR. The JARs for every core InterMine data source are published in JCenter. The build looks for jar with the name matching “bio-source-&lt;source-type&gt;-<version>.jar”, e.g. bio-source-uniprot-2.0.0.jar. Maven will automatically download the correct JARs for you.</li>
<li>The UniProt data parser reads the original XML and creates Items which are metadata representations of the objects that will be loaded into the biotestmine database. These items are stored in an intermediate items database (more about Items later).</li>
<li>Reads from the items database, converts items to objects and loads them into the biotestmine database.</li>
</ol>
<p>This should complete after a couple of minutes. Now that the data has
loaded, log into the database and view the contents of the protein
table:</p>
<hr>
<p><code>$ psql biotestmine</code>
<code>biotestmine# select count(*) from protein;</code></p>
<hr>
<p>And see the first few rows of data:</p>
<hr>
<p><code>biotestmine# select * from protein limit 5;</code></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="object-relational-mapping"></a><a href="#object-relational-mapping" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Object relational mapping</h3>
<p>InterMine works with objects, objects are loaded into the production system and queries return lists of objects. These objects are persisted to a relational database. Internal InterMine code (the ObjectStore) handles the storage and retrieval of objects from the database automatically. By using an object model InterMine queries benefit from inheritance, for example the <strong><em>Gene</em></strong> and <strong><em>Exon</em></strong> classes are both subclasses of <strong><em>SequenceFeature</em></strong>. When querying for SequenceFeatures (representing any genome feature) both Genes and Exons will be returned automatically.</p>
<p>We can see how see how inheritance is represented in the database:</p>
<ul>
<li>One table is created for each class in the data model.</li>
<li>Where one class inherits from another, entries are written to both
tables. For example:</li>
</ul>
<hr>
<p><code>biotestmine# select * from gene limit 5;</code></p>
<hr>
<p>The same rows appear in the sequencefeature table:</p>
<hr>
<p><code>biotestmine# select * from sequencefeature limit 5;</code></p>
<hr>
<p>All classes in the object model inherit from InterMineObject. Querying the intermineobject table in the database is a useful way to find the total number of objects in a Mine:</p>
<hr>
<p><code>biotestmine# select count(*) from intermineobject;</code></p>
<hr>
<p>All tables include an id column for unique ids and a class column with the actual class of that object. Querying the class column of intermineobject you can find the counts of different objects in a Mine:</p>
<hr>
<p><code>biotestmine# select class, count(*) from intermineobject group by class;</code></p>
<hr>
<p>A technical detail: for speed when retrieving objects and to deal with inheritance correctly (e.g. to ensure a Gene object with all of its fields is returned even if the query was on the <strong><em>SequenceFeature</em></strong> class) a serialised copy of each object is stored in the <strong><em>intermineobject</em></strong> table. When queries are run by the ObjectStore they actually return the ids of objects - these objects are may already be in a cache, if not the are retrieved from the intermineobject table.</p>
<h3><a class="anchor" aria-hidden="true" id="loading-genome-data-from-gff3-and-fasta"></a><a href="#loading-genome-data-from-gff3-and-fasta" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Loading Genome Data from GFF3 and FASTA</h3>
<p>We will load genome annotation data for <em>P. falciparum</em> from PlasmoDB</p>
<ul>
<li>genes, mRNAs, exons and their chromosome locations - in GFF3 format</li>
<li>chromosome sequences - in FASTA format</li>
</ul>
<h4><a class="anchor" aria-hidden="true" id="data-integration"></a><a href="#data-integration" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Data integration</h4>
<p>Note that genes from the gff3 file will have the same primaryIdentifier as those already loaded from UniProt. These will merge in the database such that there is only one copy of each gene with information from both data sources. We will load the genome data then look at how data integration in InterMine works.</p>
<p>First, look at the information currently loaded for gene PFL1385c from UniProt:</p>
<hr>
<p><code>biotestmine=# select * from gene where primaryIdentifier = 'PFL1385c';</code></p>
<hr>
<h4><a class="anchor" aria-hidden="true" id="gff3-files"></a><a href="#gff3-files" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>GFF3 files</h4>
<p><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3</a>
is a standard format use to represent genome features and their
locations, each line represents one feature and has nine tab-delimited
columns:</p>
<hr>
<pre><code class="hljs">MAL1    ApiDB   gene    <span class="hljs-number">183057</span>  <span class="hljs-number">184457</span>  .       -       .       ID=gene<span class="hljs-number">.46311</span>;description=hypothetical%<span class="hljs-number">20</span>protein;Name=PFA0210c
MAL1    ApiDB   mRNA    <span class="hljs-number">183057</span>  <span class="hljs-number">184457</span>  .       +       .       ID=mRNA<span class="hljs-number">.46312</span>;Parent=gene<span class="hljs-number">.46311</span>
MAL1    ApiDB   exon    <span class="hljs-number">183057</span>  <span class="hljs-number">184457</span>  .       -       <span class="hljs-number">0</span>       ID=exon<span class="hljs-number">.46313</span>;Parent=mRNA<span class="hljs-number">.46312</span>
</code></pre>
<hr>
<p><strong>col 1: “seqid”</strong>
:  an identifier for a ‘landmark’ on which the current feature is locatated, in this case ‘MAL1’, a ‘’P. falciparum’’ chromosome.</p>
<p><strong>col 2: “source”</strong>
:    the database or algorithm that provided the feature</p>
<p><strong>col 3: “type”</strong>
:    a valid Sequence Ontology term defining the feature type - here gene or mRNA</p>
<p><strong>col 4 &amp; 5: “start” and “end”</strong>
:    coordinates of the feature on the landmark in col 1</p>
<p><strong>col 6: “score”</strong>
:   an optional score, used if the feature has been generated by an algorithm</p>
<p><strong>col 7: “strand”</strong>
:   ‘+’ or ‘-‘ to indicate the strand the feature is on</p>
<p><strong>col 8: “phase”</strong>
:    for CDS features to show where the feature begins with reference to the reading frame</p>
<p><strong>col 9: “attributes”</strong>
:  custom attributes to describe the feature, these are name/value pairs separated by ‘;’. Some attributes have predefined meanings, relevant here:</p>
<ul>
<li>ID - identifier of feature, unique in scope of the GFF3 file</li>
<li>Name - a display name for the feature</li>
<li>Parent - the ID of another feature in the file that is a parent of this one. In our example the gene is a Parent of the mRNA.</li>
</ul>
<p>A dot means there is no value provided for the column.
The files we are loading are from PlasmoDB and contain gene, exon and mRNA features, there is one file per chromosome. Look at an example:</p>
<hr>
<p><code>$ less /data/malaria/genome/gff/MAL1.gff3</code></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="the-gff3-source"></a><a href="#the-gff3-source" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>The GFF3 source</h3>
<p>InterMine includes a parser to load valid GFF3 files. The creation of features, sequence features, locations and standard attributes is taken care of automatically.</p>
<p>Other gff3 properties can be configured in the project.xml The properties set for malaria-gff are:</p>
<p>gff3.seqClsName = Chromosome</p>
<p>:   the ids in the first column represent Chromosome objects, e.g. MAL1</p>
<p>gff3.taxonId = 36329</p>
<p>:   taxon id of malaria</p>
<p>gff3.dataSourceName = PlasmoDB</p>
<p>:   the data source for features and their identifiers, this is used for
the DataSet (evidence) and synonyms.</p>
<p>gff3.seqDataSourceName = PlasmoDB</p>
<p>:   the source of the seqids (chromosomes) is sometimes different to the
features described</p>
<p>gff3.dataSetTitle = PlasmoDB P. falciparum genome</p>
<p>:   a DataSet object is created as evidence for the features, it is
linked to a DataSource (PlasmoDB)</p>
<p>You can also configure GFF properties in the gff.config file. See <a href="'https://intermine.readthedocs.io/en/latest/database/data-sources/library/gff/'">GFF3</a> for details.</p>
<p>To deal with any specific attributes or perform custom operations on each feature you can write a handler in Java which will get called when reading each line of GFF. For malaria gff we need a handler to switch which fields from the file are set as primaryIdentifier and symbol/secondaryIdentifier in the features created. This is to match the identifiers from UniProt, it is quite a common issue when integrating from multiple data sources.
To deal with any specific attributes or perform custom operations on each feature you can write a handler in Java which will get called when reading each line of GFF. For malaria gff we need a handler to switch which fields from the file are set as primaryIdentifier and symbol/secondaryIdentifier in the features created. This is to match the identifiers from UniProt, it is quite a common issue when integrating from multiple data sources.</p>
<p>From the example above, by default: ID=gene.46311;description=hypothetical%20protein;Name=PFA0210c would make Gene.primaryIdentifier be gene.46311 and Gene.symbol be PFA0210c. We need PFA0210c to be the primaryIdentifier.</p>
<p>Look at the malaria-gff.properties file - there are two properties of interest:</p>
<hr>
<!--set the source type to be gff-->
`have.file.gff=true`
<!---specify a Java class to be called on each row of the gff file to cope with attributes-->
`gff3.handlerClassName = org.intermine.bio.dataconversion.MalariaGFF3RecordHandler`
<hr>
<p>The property file has specified a Java class to process the GFF file,
<a href="https://github.com/intermine/intermine/blob/master/bio/sources/example-sources/malaria-gff/src/main/java/org/intermine/bio/dataconversion/MalariaGFF3RecordHandler.java">MalariaGFF3RecordHandler</a>. This code changes which fields the [ID]{.title-ref} and
[Name]{.title-ref} attributes from the GFF file have been assigned to.</p>
<h3><a class="anchor" aria-hidden="true" id="loading-gff3-data"></a><a href="#loading-gff3-data" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Loading GFF3 data</h3>
<p>Now execute the [malaria-gff]{.title-ref} source by running this
command:</p>
<hr>
<pre><code class="hljs">&lt;!---load the GFF data--&gt;
</code></pre>
<p><code>~/git/biotestmine $ ./gradlew integrate -Psource=malaria-gff --stacktrace</code></p>
<hr>
<p>This will take a few minutes to run. Note that this time we don't run
[buildDB]{.title-ref} as we are loading this data into the same database
as UniProt. As before you can run a query to see how many objects of
each class are loaded:</p>
<hr>
<p><code>$ psql biotestmine</code>
<code>biotestmine# select class, count(*) from intermineobject group by class;</code></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="fasta-files"></a><a href="#fasta-files" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>FASTA files</h3>
<p>FASTA is a minimal format for representing sequence data. Files comprise a header with some identifier information preceded by ‘&gt;’ and a sequence. At present the InterMine FASTA parser loads just the first entry in header after &gt; and assigns it to be an attribute of the feature created. Here we will load one FASTA file for each malaria chromosome. Look at an example of the files we will load:</p>
<hr>
<p><code>$ less /data/malaria/genome/fasta/MAL1.fasta</code></p>
<hr>
<p>The type of feature created is defined by a property in project.xml, the attribute set defaults to primaryIdentifier but can be changed with the fasta.classAttribute property. The following properties are defined in project.xml for malaria-chromosome-fasta:</p>
<p>fasta.className = org.intermine.model.bio.Chromosome
:    the type of feature that each sequence is for</p>
<p>fasta.dataSourceName = PlasmoDB
:    the source of identifiers to be created</p>
<p>fasta.dataSetTitle = PlasmoDB chromosome sequence
:    a DataSet object is created as evidence</p>
<p>fasta.taxonId = 36329
:    the organism id for malaria</p>
<p>fasta.includes = MAL*.fasta
:    files to process</p>
<ul>
<li></li>
</ul>
<p>This will create features of the class Chromosome with primaryIdentifier set and the Chromosome.sequence reference set to a Sequence object. Also created are a DataSet and DataSource as evidence.</p>
<h3><a class="anchor" aria-hidden="true" id="loading-fasta-data"></a><a href="#loading-fasta-data" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Loading FASTA data</h3>
<p>Now run the malaria-chromosome-fasta source by running this command:</p>
<hr>
<pre><code class="hljs">&lt;!---load FASTA data--&gt;
</code></pre>
<p><code>~/git/biotestmine $ ./gradlew integrate -Psource=malaria-chromosome-fasta --stacktrace</code></p>
<hr>
<p>This has integrated the chromosome objects with those already in the database. In the next step we will look at how this data integration works.</p>
<h3><a class="anchor" aria-hidden="true" id="data-integration-1"></a><a href="#data-integration-1" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Data Integration</h3>
<h4><a class="anchor" aria-hidden="true" id="data-integration-in-biotestmine"></a><a href="#data-integration-in-biotestmine" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Data integration in BioTestMine</h4>
<p>The sources uniprot-malaria and malaria-gff have both loaded information about the same genes. Before loading genome data we ran a query to look at the information UniProt provided about the gene “PFL1385c”:</p>
<hr>
<p><code>biotestmine=</code><!--- select id, primaryidentifier, secondaryidentifier, symbol, length , chromosomeid, chromosomelocationid, organismid from gene where primaryIdentifier = 'PFL1385c';--></p>
<table>
<thead>
<tr><th>id</th><th>primaryidentifier</th><th>secondaryidentifier</th><th>symbol</th><th>length</th><th>chromosomeid</th><th>chromosomelocationid</th><th>organismid</th></tr>
</thead>
<tbody>
<tr><td>83000626</td><td>PFL1385c</td><td></td><td>ABRA</td><td></td><td></td><td></td><td>83000003</td></tr>
</tbody>
</table>
<hr>
<p>Which showed that UniProt provided primaryIdentifier and symbol attributes and set the organism reference. The id was set automatically by the ObjectStore and will be different each time you build your Mine.</p>
<p>Running the same query after malaria-gff is added shows that more fields have been filled in for same gene and that it has kept the same id:</p>
<hr>
<p><code>biotestmine=</code><!--- select id, primaryidentifier, secondaryidentifier, symbol, length , chromosomeid, chromosomelocationid, organismid from gene where primaryIdentifier = 'PFL1385c';--></p>
<table>
<thead>
<tr><th>id</th><th>primaryidentifier</th><th>secondaryidentifier</th><th>symbol</th><th>length</th><th>chromosomeid</th><th>chromosomelocationid</th><th>organismid</th></tr>
</thead>
<tbody>
<tr><td>83000626</td><td>PFL1385c</td><td>gene.33449</td><td>ABRA</td><td>2232</td><td>84017653</td><td>84018828</td><td>83000003</td></tr>
</tbody>
</table>
<hr>
<p>This means that when the second source was loaded the integration code was able to identify that an equivalent gene already existed and merged the values for each source, the equivalence was based on primaryIdentifier as this was the field that the two sources had in common.</p>
<p>Note that malaria-gff does not include a value for symbol but it did not write over the symbol provided by UniProt, actual values always take precedence over null values (unless configured otherwise).
Now look at the organism table:</p>
<hr>
<p><code>biotestmine=</code># select * from organism;</p>
<table>
<thead>
<tr><th>genus</th><th>taxonid</th><th>species</th><th>abbreviation</th><th>id</th><th>shortname</th><th>name</th><th>class</th></tr>
</thead>
<tbody>
<tr><td></td><td>36329</td><td></td><td></td><td>83000003</td><td></td><td></td><td>org.intermine.model.genomic.Organism</td></tr>
</tbody>
</table>
<hr>
<p>Three sources have been loaded so far that all included the organism with taxonId 36329, and more importantly they included objects that reference the organism. There is still only one row in the organism table so the data from three sources has merged, in this case taxonId was the field used to define equivalence.</p>
<h4><a class="anchor" aria-hidden="true" id="how-data-integration-works"></a><a href="#how-data-integration-works" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>How data integration works</h4>
<p>Data integration works by defining keys for each class of object to describe fields that can be used to define equivalence for objects of that class. For the examples above:</p>
<ul>
<li>primaryIdentifier was used as a key for Gene</li>
<li>taxonId was used as a key for Organism</li>
</ul>
<p>For each Gene object loaded by malaria-gff a query was performed in the biotestmine database to find any existing Gene objects with the same primaryIdentifier. If any were found fields from both objects were merged and the resulting object stored.</p>
<p>Many performance optimisation steps are applied to this process. We don’t actually run a query for each object loaded, requests are batched and queries can be avoided completely if the system can work out no integration will be needed.</p>
<p>We may also load data from some other source that provides information about genes but doesn’t use the identifier scheme we have chosen for primaryIdentifier (in our example PFL1385c). Instead it only knows about the symbol (ABRA), in that case we would want that source to use the symbol to define equivalence for Gene.
Important points:</p>
<ul>
<li>A key defines a field or fields of a class that can be used to search for equivalent objects</li>
<li>Multiple primary keys can be defined for a class, sources can use different keys for a class if they provide different identifiers</li>
<li>One source can use multiple primary keys for a class if the objects of that class don’t consistently have the same identifier type</li>
<li>null - if a source has no value for a field that is defined as a primary key then the key is not used and the data is loaded without being integrated.</li>
</ul>
<h4><a class="anchor" aria-hidden="true" id="integration-keys-in-biotestmine"></a><a href="#integration-keys-in-biotestmine" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Integration Keys in BioTestMine</h4>
<p>The keys used by each source are set in the source’s resources directory.</p>
<ul>
<li><a href="https://github.com/intermine/intermine/blob/master/bio/sources/uniprot/src/main/resources/uniprot_keys.properties">uniprot-malaria</a></li>
<li><a href="https://github.com/intermine/intermine/blob/master/bio/sources/example-sources/malaria-gff/src/main/resources/malaria-gff_keys.properties">malaria-gff</a></li>
</ul>
<p>The key on Gene.primaryIdentifier is defined in both sources, that means that the same final result would have been achieved regardless of the order in the two sources were loaded.</p>
<p>These _keys.properties files define keys in the format:</p>
<hr>
<p><code>Class.name_of_key = field1, field2</code></p>
<hr>
<p>The name_of_key can be any string but you must use different names if defining more than one key for the same class, for example in uniprot_keys.properties there are two different keys defined for Gene:</p>
<hr>
<p><code>Gene.key_primaryidentifier = primaryIdentifier</code>
<code>Gene.key_secondaryidentifier = secondaryIdentifier</code></p>
<hr>
<p>It is better to use common names for identical keys between sources as this will help avoid duplicating database indexes. Each key should list one or more fields that can be a combination of attributes of the class specified or references to other classes, in this cases there should usually be a key defined for the referenced class as well.</p>
<h4><a class="anchor" aria-hidden="true" id="the-tracker-table"></a><a href="#the-tracker-table" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>The tracker table</h4>
<p>A special tracker table is created in the target database by the data integration system. This tracks which sources have loaded data for each field of each object. The data is used along with priorities configuration when merging objects but is also useful to view where objects have come from.</p>
<ul>
<li>Look at the columns in the tracker table, objectid references an object from some other table</li>
<li>Query tracker information for the objects in the examples above:</li>
</ul>
<hr>
<p><code>select distinct sourcename from tracker, gene where objectid = id and primaryidentifier = 'PFL1385c';</code></p>
<p><code>select objectid, sourcename, fieldname, version from tracker, gene where objectid = id and primaryidentifier = 'PFL1385c';</code></p>
<p><code>select distinct sourcename from tracker, organism where objectid = id;</code></p>
<hr>
<h4><a class="anchor" aria-hidden="true" id="updating-organism-and-publication-information"></a><a href="#updating-organism-and-publication-information" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Updating Organism and Publication Information</h4>
<p>Organisms and publications in InterMine are loaded by their taxon id and PubMed id respectively. The entrez-organism and update-publications sources can be run at the end of the build to examine the ids loaded, fetch details via the NCBI Entrez web service and add those details to the Mine.</p>
<h4><a class="anchor" aria-hidden="true" id="fetching-organism-details"></a><a href="#fetching-organism-details" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Fetching organism details</h4>
<p>You will have noticed that in previous sources and in project.xml we have referred to organisms by their NCBI Taxonomy id. These are numerical ids assigned to each species. We use these for convenience in integrating data, the taxon id is a good unique identifier for organisms whereas names can come in many different formats: for example in fly data sources we see: Drosophila melanogaster, D. melanogaster, Dmel, DM, etc.</p>
<p>Looking at the organism table in the database you will see that the only column filled in is taxonid:</p>
<hr>
<p><code>$ psql biotestmine</code>
<code>biotestmine# select * from organism;</code></p>
<hr>
<p>From the root biotestmine directory run the entrez-organism source:</p>
<hr>
<!--- load organism data -->
`~/git/biotestmine $ ./gradlew integrate -Psource=entrez-organism --stacktrace`
___
This should only take a few seconds. This source does the following:
<ul>
<li>runs a query in the production database for all of the taxon ids</li>
<li>creates an NCBI Entrez web service request to fetch details of those
organisms</li>
<li>converts the data returned from Entrez into a temporary Items XML
file</li>
<li>loads the Items XML file into the production database</li>
</ul>
<p>Now run the same query in the production database, you should see details for ‘’P. falciparum’’ added:</p>
<hr>
<p><code>$ psql biotestmine</code>
<code>biotestmine# select * from organism;</code></p>
<hr>
<p>As this source depends on organism data previously loaded it should be one of the last sources run and should appear at the end of <sources> in project.xml.</p>
<h4><a class="anchor" aria-hidden="true" id="fetching-publication-details"></a><a href="#fetching-publication-details" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Fetching publication details</h4>
<p>Publications are even more likely to be cited in different formats and
are prone to errors in their description. We will often load data
referring to the same publication from multiple sources and need to
ensure those publications are integrated correctly. Hence we load only
the PubMed id and fetch the details from the NCBI Entrez web service as
above.</p>
<p>Several InterMine sources load publications:</p>
<hr>
<p><code>biotestmine# select count(*) from publication;</code>
<code>biotestmine# select * from publication limit 5;</code></p>
<hr>
<p>Now details will have been added to the publication table:</p>
<hr>
<p><code>~/git/biotestmine $ ./gradlew integrate -Psource=update-publications --stacktrace</code></p>
<hr>
<p>As there are often large numbers of publications they are retrieved in
batches from the web service.</p>
<p>Now details will have been added to the [publication]{.title-ref} table:</p>
<hr>
<p><code>biotestmine# select * from publication where title is not null limit 5;</code></p>
<hr>
<p>As this source depends on publication data previously loaded it should be one of the last sources run and should appear at the end of <sources> in project.xml.</p>
<h4><a class="anchor" aria-hidden="true" id="post-processing"></a><a href="#post-processing" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Post Processing</h4>
<p>Post-processing steps are run after all data is loaded, they are specified as &lt;post-process&gt; elements in project.xml.</p>
<p>Some of these can only be run after data from multiple sources are loaded. For example, for the Malaria genome information we load features and their locations on chromosomes from malaria-gff but the sequences of chromosomes from malaria-chromosome-fasta. These are loaded independently and the Chromosome objects from each are integrated, neither of these on their own could set the sequence of each Exon. However, now they are both loaded the transfer-sequences post-process can calculate and set the sequences for all features located on a Chromosome for which the sequence is known.</p>
<p>Some post-process steps are used to homogenize data from different sources or fill in shortcuts in the data model to improve usability - e.g. create-references.</p>
<p>Finally, there are post-process operations that create summary information to be used by the web application: summarise-objectstore, create-search-index and create-autocomplete-indexes.</p>
<h4><a class="anchor" aria-hidden="true" id="biotestmine-post-processing"></a><a href="#biotestmine-post-processing" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>BioTestMine Post Processing</h4>
<p>The following &lt;post-process&gt; targets are included in the BioTestMine project.xml.</p>
<p>Run queries listed here before and after running the post-processing to see examples of what each step does.</p>
<h5><a class="anchor" aria-hidden="true" id="create-references"></a><a href="#create-references" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>create-references</h5>
<p>This fills in some shortcut references in the data model to make querying easier. For example, Gene has a collection of transcripts and Transcript has a collection of exons. create-references will follow these collections and create a gene reference in Exon and the corresponding exons collection in Gene.</p>
<hr>
<p><code>biotestmine# select * from exon limit 5;</code></p>
<hr>
<p>The empty geneid column will be filled in representing the reference to gene.</p>
<p>Execute the create-references postprocess by running this command:</p>
<hr>
<!-- execute create-references postprocess-->
` ~/git/biotestmine $ ./gradlew postprocess -Pprocess=create-references`
___
##### transfer-sequences
The sequence for chromosomes is loaded by malaria-chromosome-fasta but no sequence is set for the features located on them. This step reads the locations of features, calculates and stores their sequence and sets the sequenceid column. The sequenceid column for this exon is empty:
___
biotestmine <!---select * from exon where primaryidentifier = 'exon.32017';-->
___
Execute the transfer-sequences postprocess by running this command:
___
<!--- execute transfer-sequences postprocess-->
`~/git/biotestmine $ ./gradlew postprocess -Pprocess=transfer-sequences`
____
After running transfer-sequences the sequenceid column is filled in.
<h5><a class="anchor" aria-hidden="true" id="do-sources"></a><a href="#do-sources" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>do-sources</h5>
<p>Each source can also provide code to execute post-process steps if required. This command loops through all of the sources and checks whether there are any post-processing steps configured. There aren’t any for the sources we are using for BioTestMine but you should always include the do-sources element.</p>
<h5><a class="anchor" aria-hidden="true" id="summarise-objectstore-create-search-index--create-autocomplete-index"></a><a href="#summarise-objectstore-create-search-index--create-autocomplete-index" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>summarise-objectstore, create-search-index &amp; create-autocomplete-index</h5>
<p>These generate summary data and search indexes used by the web application, see <a href="'https://intermine.readthedocs.io/en/latest/webapp/keyword-search/'">Keyword Search</a> for details.</p>
<p>Execute the summarise-objectstore postprocess by running this command:</p>
<hr>
<pre><code class="hljs">&lt;!--- execute transfer-sequences postprocess--&gt;
</code></pre>
<p><code>~/git/biotestmine $ ./gradlew postprocess -Pprocess=summarise-objectstore</code></p>
<hr>
<p>You must have Solr installed and running for the indexes to be populated
correctly.</p>
<p><strong>Install SOLR</strong></p>
<p>Download <a href="http://archive.apache.org/dist/lucene/solr/7.2.1/">Solr binary package</a> and extract
it to any place you like.Inside /solr-7.2.1 directory start the server with this command:</p>
<hr>
   <!-- Starts the server instance on port 8983 -->
   `solr-7.2.1 $ ./bin/solr start`
___
**Initialising Search Indexes**
<p>To create a Intermine collection for search process, run this command
inside the solr directory.</p>
<hr>
<!--- Initialises the search index-->
`    solr-7.2.1 $ ./bin/solr create -c biotestmine-search`
____
To create a Intermine collection for autocomplete process, run this
command inside the solr directory.
<!--- Initaliases the autocomplete index-->
`    solr-7.2.1 $ ./bin/solr create -c biotestmine-autocomplete`
<p>These are empty search indexes. These will be populated by the create-search-index &amp; create-autocomplete-index postprocesses.</p>
<p>See <a href="'https://intermine.readthedocs.io/en/latest/system-requirements/software/solr/'">Solr</a> for details.</p>
<p>Execute the create-search-index and create-autocomplete-index postprocesses by running this command:</p>
<hr>
<!--- execute create-search-index and create-autocomplete-index postprocesse-->
`    ~/git/biotestmine $ ./gradlew postprocess -Pprocess=create-search-index`
`   ~/git/biotestmine $ ./gradlewpostprocess-Pprocess=create-autocomplete-index`
___
#### Building a Mine with a Perl script
<p>So far we have created databases, integrated data and run post-processing with individual gradle tasks. Alternatively InterMine has a Perl program called project_build that reads the project.xml definition and runs all of the steps in sequence. The script has the option of creating snapshots during the build at specified checkpoints.</p>
<h5><a class="anchor" aria-hidden="true" id="build-complete-biotestmine"></a><a href="#build-complete-biotestmine" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Build complete BioTestMine</h5>
<p>To build BioTestMine using the project_build script, first download the script:</p>
<hr>
<!--- download the script-->
`  ~/git/biotestmine $ wget https://raw.githubusercontent.com/intermine/intermine-scripts/master/project_build`
<!--- make executable-->
`    ~/git/biotestmine $ chmod +x project_build`
_____
Run the project_build script from your biotestmine directory:
_____
`    ~/git/biotestmine $ ./project_build -b -v localhost ~/biotestmine-dump`
____
This will take ~15-30mins to complete.
-
---
**NOTE**
<p>If you encounter an “OutOfMemoryError”, you should set your $GRADLE_OPTS variable, see <a href="'https://intermine.readthedocs.io/en/latest/support/troubleshooting-tips/'">Troubleshooting tips</a></p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="deploying-the-web-application"></a><a href="#deploying-the-web-application" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Deploying the web application</h3>
<p>You can deploy a web application against your newly built database.</p>
<h3><a class="anchor" aria-hidden="true" id="configure"></a><a href="#configure" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Configure</h3>
<p>In the ~/.intermine directory, update the webapp properties in your biotestmine.properties file. Update the following properties:</p>
<ul>
<li>tomcat username and password</li>
<li>superuser username and password</li>
</ul>
<h3><a class="anchor" aria-hidden="true" id="userprofile"></a><a href="#userprofile" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>UserProfile</h3>
<p>The userprofile database stores all user-related information such as
username and password, tags, queries, lists and templates.</p>
<ol>
<li>Configure
Update your biotestmine.properties file with correct information for the db.userprofile-production database:</li>
</ol>
<hr>
<p><code>db.userprofile-production.datasource.serverName=DB_SERVER</code>
<code>db.userprofile-production.datasource.databaseName=userprofile-biotestmine</code>
<code>db.userprofile-production.datasource.user=USER_NAME</code>
<code>db.userprofile-production.datasource.password=USER_PASSWORD</code></p>
<hr>
<ol start="2">
<li>Create the empty database:</li>
</ol>
<hr>
<p><code>$ createdb userprofile-biotestmine</code></p>
<hr>
<ol start="3">
<li>Build the database:</li>
</ol>
<hr>
<!--- creates the empty tables-->
`    ~/git/biotestmine $ ./gradlew buildUserDB`
_____
You only need to build the userprofile database once.
<hr>
<p><strong>Warning</strong>
The buildDB and buildUserDB commands rebuild the database and thus will delete any data.</p>
<hr>
<h3><a class="anchor" aria-hidden="true" id="deploying-the-webapp"></a><a href="#deploying-the-webapp" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Deploying the webapp</h3>
<p>Before deploying the biotestmine webapp, you need to configure tomcat. See <a href="'https://intermine.readthedocs.io/en/latest/system-requirements/software/tomcat/'">Tomcat</a> for configuration details.</p>
<p>Run the following command to release your webapp:</p>
<hr>
<!--- deploy the webapp (tomcat must be running) -->
`    ~/git/biotestmine $ ./gradlew cargoDeployRemote`
____
If you make changes, redeploy your webapp with this command:
_____
<!--- REdeploy the webapp (tomcat must be running) -->
`    ~/git/biotestmine $ ./gradlew cargoReDeployRemote`
____
### Using the webapp
<p>Navigate to <a href="'http://localhost:8080/biotestmine'">http://localhost:8080/biotestmine </a> to view your webapp. The path to your webapp is the webapp.path value set in biotestmine.properties.</p>
<hr>
<h2><a class="anchor" aria-hidden="true" id="next"></a><a href="#next" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a><strong>Next</strong></h2>
<p>Now that you have a database and a working webapp, you’ll want to know how to add your own logo, pick a colour scheme, modify how data is displayed etc. Our <a href="'https://intermine.readthedocs.io/en/latest/get-started/tutorial/webapp/'">webapp tutorial</a> is a detailed guide on how to customise all parts of the InterMine web application..</p>
<h3><a class="anchor" aria-hidden="true" id="help"></a><a href="#help" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Help</h3>
<h4><a class="anchor" aria-hidden="true" id="gradle"></a><a href="#gradle" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Gradle</h4>
<p>Anytime you run <strong><em>./gradlew</em></strong> and something bad happens, add the <strong><em>–stacktrace</em></strong> or <strong><em>–debug</em></strong> options.</p>
<p>This will give you more detailed output and hopefully a more helpful error message.</p>
<h3><a class="anchor" aria-hidden="true" id="logs"></a><a href="#logs" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Logs</h3>
<p>If the error occurs while you are integrating data, the error message will be in the intermine.log file in the directory you are in.</p>
<p>If the error occurs while you are browsing your webapp, the error message will be located in the Tomcat logs: $TOMCAT/logs.</p>
<h3><a class="anchor" aria-hidden="true" id="contact-us"></a><a href="#contact-us" aria-hidden="true" class="hash-link"><svg class="hash-link-icon" aria-hidden="true" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Contact us!</h3>
<p>Please <a href="http://intermine.org/contact/">contact us</a> if you run into
problems. We have a discord server, twitter and a developer mailing
list.</p>
</span></div></article></div><div class="docs-prevnext"><a class="docs-prev button" href="/test-site/docs/next/doc1"><span class="arrow-prev">← </span><span>Example Page</span></a><a class="docs-next button" href="/test-site/docs/next/doc2"><span>document number 2</span><span class="arrow-next"> →</span></a></div></div></div><nav class="onPageNav"><ul class="toc-headings"><li><a href="#software">Software</a></li><li><a href="#biotestmine">BioTestMine</a><ul class="toc-headings"><li><a href="#get-rid-of-daemons">Get rid of daemons</a></li><li><a href="#help-somethings-gone-wrong">Help! Something's gone wrong.</a></li></ul></li><li><a href="#projectxml">Project.xml</a><ul class="toc-headings"><li><a href="#properties-file">Properties file</a></li><li><a href="#create-databases">Create databases</a></li></ul></li><li><a href="#the-data-model">The Data Model</a><ul class="toc-headings"><li><a href="#defining-the-model">Defining the model</a></li><li><a href="#corexml">core.xml</a></li><li><a href="#sequence-ontology">Sequence Ontology</a></li><li><a href="#additions-files">Additions files</a></li><li><a href="#creating-a-database">Creating a database</a></li><li><a href="#loading-data">Loading Data</a></li><li><a href="#loading-data-from-a-source">Loading data from a source</a></li><li><a href="#object-relational-mapping">Object relational mapping</a></li><li><a href="#loading-genome-data-from-gff3-and-fasta">Loading Genome Data from GFF3 and FASTA</a></li><li><a href="#the-gff3-source">The GFF3 source</a></li><li><a href="#loading-gff3-data">Loading GFF3 data</a></li><li><a href="#fasta-files">FASTA files</a></li><li><a href="#loading-fasta-data">Loading FASTA data</a></li><li><a href="#data-integration-1">Data Integration</a></li></ul></li><li><a href="#this-will-take-15-30mins-to-complete">This will take ~15-30mins to complete.</a><ul class="toc-headings"><li><a href="#deploying-the-web-application">Deploying the web application</a></li><li><a href="#configure">Configure</a></li><li><a href="#userprofile">UserProfile</a></li><li><a href="#deploying-the-webapp">Deploying the webapp</a></li><li><a href="#using-the-webapp">Using the webapp</a></li></ul></li><li><a href="#next"><strong>Next</strong></a><ul class="toc-headings"><li><a href="#help">Help</a></li><li><a href="#logs">Logs</a></li><li><a href="#contact-us">Contact us!</a></li></ul></li></ul></nav></div><footer class="nav-footer" id="footer"><section class="sitemap"><a href="/test-site/" class="nav-home"><img src="/test-site/img/favicon.ico" alt="Test Site" width="66" height="58"/></a><div><h5>Docs</h5><a href="/test-site/docs/doc1.html">Getting Started (or other categories)</a><a href="/test-site/docs/doc2.html">Guides (or other categories)</a><a href="/test-site/docs/doc3.html">API Reference (or other categories)</a></div><div><h5>Community</h5><a href="/test-site/users">User Showcase</a><a href="https://stackoverflow.com/questions/tagged/" target="_blank" rel="noreferrer noopener">Stack Overflow</a><a href="https://discordapp.com/">Project Chat</a><a href="https://twitter.com/" target="_blank" rel="noreferrer noopener">Twitter</a></div><div><h5>More</h5><a href="/test-site/blog">Blog</a><a href="https://github.com/">GitHub</a><a class="github-button" data-icon="octicon-star" data-count-href="/facebook/docusaurus/stargazers" data-show-count="true" data-count-aria-label="# stargazers on GitHub" aria-label="Star this project on GitHub">Star</a></div></section><a href="https://opensource.facebook.com/" target="_blank" rel="noreferrer noopener" class="fbOpenSource"><img src="/test-site/img/oss_logo.png" alt="Facebook Open Source" width="170" height="45"/></a><section class="copyright">Copyright © 2020 Your Name or Your Company Name</section></footer></div></body></html>